File: tests/e2e/live.mjs
import assert from "node:assert/strict";
import { readFile, writeFile } from "node:fs/promises";
import { createAgent } from "@microsoft/amplifier-agent";

const resume = process.env.E2E_MODE === "resume";
const effects = [];
const normalize = value => JSON.parse(JSON.stringify(value, (_, item) =>
  typeof item === "bigint" ? String(item) : item));
const agent = await createAgent({
  tools: [{ name: "record_probe", description: "Record a harmless in-memory probe",
    inputSchema: { $schema: "https://json-schema.org/draft/2020-12/schema", type: "object",
      properties: { value: { type: "string" } }, required: ["value"] },
    handler: async arguments_ => { effects.push(arguments_); return "Probe recorded successfully."; },
  }],
  approvals: async request => ({ decision: request.name === "record_probe" ? "allow" : "deny" }),
  instructions: "Use only record_probe when asked for a tool. Keep responses short.",
});
try {
  const session = resume ? await agent.resumeSession("live-session")
    : await agent.createSession({ sessionId: "live-session" });
  if (resume) assert.deepEqual(normalize(session.history),
    JSON.parse(await readFile(process.env.E2E_EXPECTED_HISTORY, "utf8")));
  const prompt = resume
    ? "Which value did record_probe record? Reply with only that value; do not call tools."
    : "Call record_probe exactly once with value 'ok', then reply OK.";
  const turn = await session.startTurn({ content: [{ type: "text", text: prompt }] });
  const events = [];
  for await (const event of turn.events()) events.push(event);
  const result = events.at(-1).payload;
  assert.equal(result.state, "success", result.error?.code);
  assert.equal(events[0].type, "turn_started");
  assert.equal(events.at(-1).type, "terminal");
  assert.ok(events.some(event => event.type === "output_delta"));
  assert.deepEqual(session.history.at(-1).result, result);
  assert.equal(session.history.length, resume ? 2 : 1);
  assert.deepEqual(effects, resume ? [] : [{ value: "ok" }]);
  assert.ok(result.usage.entries.length);
  if (resume) {
    assert.equal(events[0].payload.continuation, "resumed");
    assert.ok(result.content.map(part => part.text).join("").toLowerCase().includes("ok"));
  } else await writeFile(process.env.E2E_EXPECTED_HISTORY, JSON.stringify(normalize(session.history)));
  await session.close();
  console.log(JSON.stringify({ kind: "result", state: result.state, streamed: true,
    effects: effects.length, history_count: resume ? 2 : 1, pid: process.pid }));
} finally { await agent.close(); }

File: tests/e2e/live.py
"""Stream a live turn through an installed binding and reopen its durable transcript."""

import asyncio
import dataclasses
import json
import os
from decimal import Decimal
from pathlib import Path

from amplifier_agent import (
    AgentOptions,
    ApprovalResponse,
    SessionOptions,
    TextPart,
    Tool,
    TurnInput,
    create_agent,
)


def normalize(value):
    if dataclasses.is_dataclass(value):
        return {key: normalize(item) for key, item in vars(value).items() if item is not None}
    if isinstance(value, dict):
        return {key: normalize(item) for key, item in value.items()}
    if isinstance(value, list):
        return [normalize(item) for item in value]
    if isinstance(value, Decimal):
        return str(value)
    return value


async def main():
    resume = os.environ["E2E_MODE"] == "resume"
    effects = []

    async def approve(request):
        return ApprovalResponse("allow" if request.name == "record_probe" else "deny")

    async def record(arguments, context):
        effects.append(arguments)
        return "Probe recorded successfully."

    options = AgentOptions(
        tools=[
            Tool(
                "record_probe",
                "Record a harmless in-memory probe",
                {
                    "$schema": "https://json-schema.org/draft/2020-12/schema",
                    "type": "object",
                    "properties": {"value": {"type": "string"}},
                    "required": ["value"],
                },
                record,
            )
        ],
        approvals=approve,
        instructions="Use only record_probe when asked for a tool. Keep responses short.",
    )
    async with await create_agent(options) as agent:
        session = (
            await agent.resume_session("live-session")
            if resume
            else await agent.create_session(SessionOptions(session_id="live-session"))
        )
        before = normalize(session.history)
        expected = Path(os.environ["E2E_EXPECTED_HISTORY"])
        if resume:
            assert before == json.loads(expected.read_text())
        prompt = (
            "Which value did record_probe record? Reply with only that value; do not call tools."
            if resume
            else "Call record_probe exactly once with value 'ok', then reply OK."
        )
        turn = await session.start_turn(TurnInput([TextPart(prompt)]))
        events = [event async for event in turn.events()]
        result = events[-1].payload
        assert result.state == "success", result.error.code if result.error else result.state
        assert events[0].type == "turn_started" and events[-1].type == "terminal"
        assert any(event.type == "output_delta" for event in events)
        assert session.history[-1].result == result
        assert len(session.history) == (2 if resume else 1)
        assert effects == ([] if resume else [{"value": "ok"}])
        assert result.usage.entries
        if resume:
            assert events[0].payload.continuation == "resumed"
            assert "ok" in "".join(part.text for part in result.content).lower()
        else:
            expected.write_text(json.dumps(normalize(session.history)))
        await session.close()
    print(
        json.dumps(
            {
                "kind": "result",
                "state": result.state,
                "streamed": True,
                "effects": len(effects),
                "history_count": 2 if resume else 1,
                "pid": os.getpid(),
            }
        ),
        flush=True,
    )


if __name__ == "__main__":
    asyncio.run(main())
